### ISSUE TITLE
"Safety Information" modal should automatically disappear when a responder enters shared chat

### ISSUE THREAD

**Issue by fabacab** [betterangels/buoy#1]
The shared chat implementation currently uses a tlk.io embedded chat, which means we cannot detect events that occur inside the chat from the app shell (because that would be XSS). [I've asked the developers of Tlkio](https://twitter.com/maymaymx/status/646204661884567552) to see if they implemented or are willing to implement `window.postMessage()` for cross-origin event handling.

This issue is blocked until they either implement this feature or we implement our own chat solution. (IIRC, @vangogh72 mentioned some level of interest in possibly doing that?)

**Comment by fabacab** [comment]
I'm researching the possibility of using webSockets to implement our own chat. I'll post code once I have something.

**Comment by fabacab** [comment]
I've got our own chat up and running on my VVV. It uses webSockets through Ratchet http://socketo.me/ and that means using composer for dependencies. So I put our code on my fork of better-angels in the BuoyChat folder, but did not include the vendor/ directory that composer builds. 

It's currently configured to run on the VVV. 

Anyways, you can see the code on my fork. I think it's all there, and you should be able to get it up and running. 

As of now, the way to specify the user name and chat room is through the URL, ie: 

http://192.168.50.4/BuoyChat/BuoyChatClient.php?name=NAME&room=ROOM

**Comment by fabacab** [comment]
Wow, to be perfectly honest I did not believe you would actually do this. Thank you for proving me wrong. I will have a look when I get to a desktop tonight and see about merging with the WP plugin.

I'll also see about restoring the repo history, it got wiped because I was careless when merging our git repo commits with WordPress.org's hosted SVN instance.

**Comment by fabacab** [comment]
Okay, so I've restored the git history and will be maintaining a separate distribution branch for pushing into the WordPress.org-hosted plugin repository. See [Repository structure](https://github.com/meitar/better-angels/wiki/Repository-structure) for a brief explanation. Hopefully, this is the last time you'll need to deal with a forced update.

I squashed your code into one commit on the `develop` branch (de76602bcfd9565547269a028820332d00383a70) for now, and it definitely needs integration work. As far as I can tell, it's only a bare-bones modification to [Ratchet's own web chat tutorial](http://socketo.me/docs/hello-world), which is cool but won't work as-is, hence why I'm leaving it on the `develop` branch for now.

Another concern I have is that Ratchet requires PHP 5.5 or later, and it comes with ~3.5MB of dependencies. It's not exactly a showstopper but feels a bit like overkill when there are far simpler WebSocket libraries for the kind of small-scale, single-server stuff we're doing in this plugin (like [Wrench](https://github.com/varspool/Wrench)). The simpler libraries also support older PHP versions.

What I'd like to do now is to go with Ratchet as you've suggested but detect whether or not we're deployed on older PHP interpreters. If so, we can fall back to the HTML we currently have using Tlk.io without losing much, especially since we don't yet have a fully integrated WebSocket implementation ourselves anyway.

Anyway, thanks for the code and apologies on the careless repository handling on my part. I'm not actually used to receiving code patches from other people so I honestly didn't expect one in this case. Again, I'm really pleased to have been proven wrong and I hope that becomes a trend.

**Comment by fabacab** [comment]
Hmm, looking into this a bit further brings up another concern: can shared hosts (y'know, the ones that run standard WordPress installs through CPanel or whatever) handle a WebSockets server in the first place? I was under the impression that a WebSockets connection can be made directly through an already-existing HTTP server, and while that does appear to be true, it doesn't seem to be a default configuration on most low-end Web hosts.

In other words, if we are going to implement a custom WebSockets chat for this tool, we need to not impose an (unreasonably difficult) additional server configuration step for laypeople. So that means we ideally would use their web host's already-existing port 80 or 443 listener or whatever is provided by the web host.

I'm not sure this can be done with WebSockets today. Can it?
